File: backend/models/user.model.js
import mongoose from "mongoose";
import bcrypt from "bcryptjs";
import crypto from "crypto";

const userSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, "Name is required"],
      trim: true,
      maxlength: [50, "Name cannot exceed 50 characters"],
    },
    email: {
      type: String,
      required: [true, "Email is required"],
      unique: true,
      lowercase: true,
      match: [/^\S+@\S+\.\S+$/, "Please enter a valid email"],
    },
    password: {
      type: String,
      required: [true, "Password is required"],
      minlength: [6, "Password must be at least 6 characters"],
      select: false,
    },
    phone: {
      type: String,
      trim: true,
    },
    avatar: {
      public_id: String,
      url: String,
    },
    role: {
      type: String,
      enum: ["user", "admin"],
      default: "user",
    },
    wishlist: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "Product",
      },
    ],
    refreshToken: { type: String, select: false },
    resetPasswordToken: { type: String, select: false },
    resetPasswordExpire: { type: Date, select: false },
    savedCard: {
      cardNumber: { type: String },
      cardholderName: { type: String },
      expiry: { type: String },
      cvv: { type: String },
      cardType: {
        type: String,
        enum: ["visa", "mastercard", "amex", "discover"],
      },
      last4: { type: String },
    },
    isEmailVerified: { type: Boolean, default: false },
    emailVerificationToken: { type: String, select: false },
    emailVerificationExpire: { type: Date, select: false },
  },
  { timestamps: true },
);

// Hash password before save
userSchema.pre("save", async function () {
  if (!this.isModified("password")) return;
  this.password = await bcrypt.hash(this.password, 12);
});

// Compare password
userSchema.methods.comparePassword = async function (candidatePassword) {
  return bcrypt.compare(candidatePassword, this.password);
};

// Generate reset token
userSchema.methods.generateResetToken = function () {
  const token = crypto.randomBytes(32).toString("hex");
  this.resetPasswordToken = crypto
    .createHash("sha256")
    .update(token)
    .digest("hex");
  this.resetPasswordExpire = Date.now() + 15 * 60 * 1000; // 15 minutes
  return token;
};

// Generate email verification token
userSchema.methods.generateEmailVerificationToken = function () {
  const token = crypto.randomBytes(32).toString("hex");
  this.emailVerificationToken = crypto
    .createHash("sha256")
    .update(token)
    .digest("hex");
  this.emailVerificationExpire = Date.now() + 24 * 60 * 60 * 1000; // 24 hours
  return token;
};

// Indexes for performance (email already indexed by unique: true)
userSchema.index({ role: 1 }); // For admin queries
userSchema.index({ refreshToken: 1 }); // For token refresh
userSchema.index({ resetPasswordToken: 1 }); // For password reset
userSchema.index({ emailVerificationToken: 1 }); // For email verification
userSchema.index({ createdAt: -1 }); // For sorting users

const User = mongoose.model("User", userSchema);

export default User;

File: mobile/lib/features/auth/data/models/user_model.dart
import '../../domain/entities/user_entity.dart';

class UserModel {
  final String id;
  final String name;
  final String email;
  final String? phone;
  final String role;
  final AvatarModel? avatar;
  final List<String> wishlist;
  final SavedCardModel? savedCard;
  final bool isEmailVerified;
  final DateTime createdAt;
  final DateTime updatedAt;

  const UserModel({
    required this.id,
    required this.name,
    required this.email,
    this.phone,
    required this.role,
    this.avatar,
    this.wishlist = const [],
    this.savedCard,
    this.isEmailVerified = false,
    required this.createdAt,
    required this.updatedAt,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      email: json['email'] as String? ?? '',
      phone: json['phone'] as String?,
      role: json['role'] as String? ?? 'user',
      avatar: json['avatar'] != null
          ? AvatarModel.fromJson(json['avatar'] as Map<String, dynamic>)
          : null,
      wishlist: (json['wishlist'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          const [],
      savedCard: json['savedCard'] != null
          ? SavedCardModel.fromJson(json['savedCard'] as Map<String, dynamic>)
          : null,
      isEmailVerified: json['isEmailVerified'] as bool? ?? false,
      createdAt: json['createdAt'] != null
          ? (DateTime.tryParse(json['createdAt'] as String)?.toLocal()) ??
              DateTime.now()
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? (DateTime.tryParse(json['updatedAt'] as String)?.toLocal()) ??
              DateTime.now()
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'email': email,
      'phone': phone,
      'role': role,
      'avatar': avatar?.toJson(),
      'wishlist': wishlist,
      'savedCard': savedCard?.toJson(),
      'isEmailVerified': isEmailVerified,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  UserEntity toEntity() {
    return UserEntity(
      id: id,
      name: name,
      email: email,
      phone: phone,
      role: role,
      avatar: avatar?.toEntity(),
      wishlist: wishlist,
      savedCard: savedCard?.toEntity(),
      isEmailVerified: isEmailVerified,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}

class AvatarModel {
  final String publicId;
  final String url;

  const AvatarModel({
    required this.publicId,
    required this.url,
  });

  factory AvatarModel.fromJson(Map<String, dynamic> json) {
    return AvatarModel(
      publicId: json['public_id'] as String? ?? '',
      url: json['url'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'public_id': publicId,
      'url': url,
    };
  }

  AvatarEntity toEntity() {
    return AvatarEntity(
      publicId: publicId,
      url: url,
    );
  }
}

class SavedCardModel {
  final String? cardNumber;
  final String cardholderName;
  final String expiry;
  final String cvv;
  final String cardType;
  final String last4;

  const SavedCardModel({
    this.cardNumber,
    required this.cardholderName,
    required this.expiry,
    required this.cvv,
    required this.cardType,
    required this.last4,
  });

  factory SavedCardModel.fromJson(Map<String, dynamic> json) {
    return SavedCardModel(
      cardNumber: json['cardNumber'] as String?,
      cardholderName: json['cardholderName'] as String? ?? '',
      expiry: json['expiry'] as String? ?? '',
      cvv: json['cvv'] as String? ?? '',
      cardType: json['cardType'] as String? ?? '',
      last4: json['last4'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'cardNumber': cardNumber,
      'cardholderName': cardholderName,
      'expiry': expiry,
      'cvv': cvv,
      'cardType': cardType,
      'last4': last4,
    };
  }

  SavedCardEntity toEntity() {
    return SavedCardEntity(
      cardNumber: cardNumber,
      cardholderName: cardholderName,
      expiry: expiry,
      cvv: cvv,
      cardType: cardType,
      last4: last4,
    );
  }
}
